crates/polars-core/src/series/arithmetic/list_borrowed.rs

```diff
@@ -1,177 +1,1026 @@
+            (BinaryOpApplyType::ListToPrimitive, Broadcast::NoBroadcast) => {},
+            _ => unreachable!(),
+        }
+
+        //
+        // General notes
+        // * Lists can be:
+        //   * Sliced, in which case the primitive/leaf array needs to be indexed starting from an
+        //     offset instead of 0.
+        //   * Masked, in which case the masked rows are permitted to have non-matching widths.
+        //
+
+        let out = match (&self.op_apply_type, &self.broadcast) {
+            (BinaryOpApplyType::ListToList, Broadcast::NoBroadcast) => {
+                let offsets_lhs = &self.data_lhs.0[0];
+                let offsets_rhs = &self.data_rhs.0[0];
+
+                assert_eq!(offsets_lhs.len_proxy(), offsets_rhs.len_proxy());
+
+                // Output primitive (and optional validity) are aligned to the LHS input.
+                let n_values = arr_lhs.len();
+                let mut out_vec: Vec<T::Native> = Vec::with_capacity(n_values);
+                let out_ptr: *mut T::Native = out_vec.as_mut_ptr();
+
+                // Counter that stops being incremented at the first row position with mismatching
+                // list lengths.
+                let mut mismatch_pos = 0;
+
+                with_match_numeric_list_op!(&self.op, self.swapped, |$OP| {
+                    for (i, ((lhs_start, lhs_len), (rhs_start, rhs_len))) in offsets_lhs
+                        .offset_and_length_iter()
+                        .zip(offsets_rhs.offset_and_length_iter())
+                        .enumerate()
+                    {
+                        if
+                            (mismatch_pos == i)
+                            & (
+                                (lhs_len == rhs_len)
+                                | unsafe { !self.outer_validity.get_bit_unchecked(i) }
+                            )
+                        {
+                            mismatch_pos += 1;
+                        }
+
+                        // Both sides are lists, we restrict the index to the min length to avoid
+                        // OOB memory access.
+                        let len: usize = lhs_len.min(rhs_len);
+
+                        for i in 0..len {
+                            let l_idx = i + lhs_start;
+                            let r_idx = i + rhs_start;
+
+                            let l = unsafe { arr_lhs.value_unchecked(l_idx) };
+                            let r = unsafe { arr_rhs.value_unchecked(r_idx) };
+                            let v = $OP(l, r);
+
+                            unsafe { out_ptr.add(l_idx).write(v) };
+                        }
+                    }
+                });
+
+                check_mismatch_pos(mismatch_pos, offsets_lhs, offsets_rhs)?;
+
+                unsafe { out_vec.set_len(n_values) };
+
+                /// Reduce monomorphization
+                #[inline(never)]
+                fn combine_validities_list_to_list_no_broadcast(
```

Comment:
idem

Reply:
I think I want to keep this here for now - it's a very specialized function that's only used in this file. If we need to use it somewhere else later I can move it then